### DISCUSSION TITLE
GeometryFeature with LineString not reliably detected in Info event

### DISCUSSION THREAD

**Discussion by jameslavery** [Mapsui/Mapsui#2237] [Q&A]
I'm adding a GeometryFeature with a LineString, which is dislaying fine. However, the GeometryFeature is not reliably being returned in the Info event when the user taps inside the feature. Sometimes the GeometryFeature is detected, more often not.

Should I be able to reliably detect tap on any part of the GeometryFeature? Or am I setting it up wrong?

Using MAUI, running on an Android emulator on a Mac.

```
    private void Map_Info(object sender, Mapsui.MapInfoEventArgs e)
    {
        Debug.WriteLine("MapInfo");
        var l = e.MapInfo?.Layer;
        var feature = e.MapInfo?.Feature;
        if (feature != null)
        {
            if (feature is GeometryFeature geometryFeature)
            {
                var name = geometryFeature["Name"];
                Debug.WriteLine($"{name}");
            }
        }
    }

    private void AddChartworkLayer()
    {
        _control.Map.Layers.Add(new WritableLayer
        {
            Name = "Chartwork",
            Style = CreateLineStringStyle(),
            IsMapInfoLayer = true
        });
    }

    public static IStyle CreateLineStringStyle()
    {
        return new VectorStyle
        {
            Fill = null,
            Outline = null,
            Line = { Color = Color.FromString("Black"), Width = 15 }
        };
    }

    private void AddPositionLine()
    {
        if (PositionLine != null)
        {
            var layers = _control.Map.Layers.FindLayer("Chartwork");
            if (layers != null)
            {
                var features = new List<IFeature>() { CreateLineFeature() };

                var layerToAddTo = layers.First() as WritableLayer;
                layerToAddTo.AddRange(features);

                _control.Map.RefreshGraphics();
            }
        }
    }

    private GeometryFeature CreateLineFeature()
    {
        var lineLength = _control.Map.Extent.Width * 0.1;
        var feature = new GeometryFeature();
        feature["Name"] = "PositionLine";
        feature.Geometry = new LineString(new[] 
        {
            new Coordinate(_control.Map.Extent.Centroid.X - lineLength/2, _control.Map.Extent.Centroid.Y - lineLength/2),
            new Coordinate(_control.Map.Extent.Centroid.X + lineLength/2, _control.Map.Extent.Centroid.Y + lineLength/2)
        }
        );
        return feature;
    }
```

**Comment by pauldendulk** [comment]
You mention 'taps inside the feature' while this is about a LineString. I would think a LineString has no 'inside', but I guess you mean it touches the visible line (which has a width, so you could say it is inside the colored part). Is this correct?

Two possible reasons:
- When moving between touch down and touch up the tap could be interpreted as a drag and there would be no event, but in that case there would be no info event at all. If it is a proper touch but misses the feature there would be an info event but no feature in it.
Mapsui used touch  detection 
- Mapsui uses color to detect if a feature is touched. If the center of your touch does not hit a pixel that is colored by the LineString there will be no touch detection. You would like to add a touch margin in this case, but that is a missing feature. It is not so easy to fix because of the way we detect touch. A workaround would be to increase the width of the line.

**Reply by jameslavery** [reply]
Hi Paul,
Correct - it's when touching the visible line, inside the coloured part.

I'm definitely getting the touch event, as I'm getting an Info event raised.

I've increased the width of the line so it's (ridiculously) wide visibly, but still get unreliable detection of the Feature. Interestingly, I don't think I'm getting the associated Layer either - but that's probably derived from the Feature if the Feature is detected?

**Reply by pauldendulk** [reply]
Then it looks like neither explanations. Do you use MapView or MapControl?

**Reply by jameslavery** [reply]
I'm using MapControl. I'll check my code to see if I can see better what's going on.

**Reply by jameslavery** [reply]
Here's "proof" that it's not detecting. The screenshot is from a physical device, with debugger tools set to show tap locations. There's a small red dot showing where I tapped (I've added the arrow to the screenshot to help find it). The console on the right (latest at the top) shows log item 53 showing MapInfo with no Feature detected. Log items 49-52 show a previous tap which **does** detect the feature

![screenshot-1699167379905](https://github.com/Mapsui/Mapsui/assets/114386324/51d54a47-92d8-4a73-86d4-88e382f36aee)

**Comment by pauldendulk** [comment]
Can you also reproduce that problem with the line on the 'Map Info' sample under 'Demo'?

![image](https://github.com/Mapsui/Mapsui/assets/963462/64b777b8-db36-4677-ae2e-88ad98cbbb96)

It could be just a matter of mouse movement. The code only allows a movement less that one pixel. 

If so, this would be a problem for polygons and points as well.

    private static bool IsClick(MPoint? currentPosition, MPoint? previousPosition)
    {
        if (currentPosition == null || previousPosition == null)
            return false;

        return
            Math.Abs(currentPosition.X - previousPosition.X) < 1 &&
            Math.Abs(currentPosition.Y - previousPosition.Y) < 1;
    }
